net: ethernet: strip vlan headers if id is zero

clamattia opened the pull request:
Allow to receive vlan frames with id zero even when the vlan interface is disabled. They are forwarded to the native interface or in other words, the vlan header is simply stripped and ignored.

https://github.com/zephyrproject-rtos/zephyr/issues/83662

maass-hamburg commented:
It might be a good idea to have this behavior behind a Kconfig and not by default. What do the others think?

clamattia commented:
> It might be a good idea to have this behavior behind a Kconfig and not by default. What do the others think?

I do not have a copy of the standard at hand. But it is my understanding, that this is the ieee 802.1q standardized behavior. Can someone confirm?

If so, I believe it should be the hard-coded option and not behind a feature flag.

clamattia commented:
This would also fix issue https://github.com/zephyrproject-rtos/zephyr/issues/83277 for me.

jukkar commented:
I could not locate the spec where this is described. The feature is called `VLAN 0 priority tagging`, some notes about this can be found in
https://en.wikipedia.org/wiki/IEEE_802.1Q
and
https://www.cisco.com/c/en/us/td/docs/switches/connectedgrid/cg-switch-sw-master/software/configuration/guide/vlan0/b_vlan_0.html

We should enable this unconditionally so no Kconfig option for this is needed.

clamattia commented:
Does someone have time to help with the tests?

jukkar commented on subsys/net/l2/ethernet/ethernet.c:
```diff
@@ -275,38 +275,44 @@ static enum net_verdict ethernet_recv(struct net_if *iface,
-			enum net_verdict verdict;
-
-			net_pkt_set_vlan_tci(pkt, ntohs(hdr_vlan->vlan.tci));
+	if (type == NET_ETH_PTYPE_VLAN) {
```
If VLAN is disabled, then we should never receive VLAN frames. So with this in mind, I think we do need to have the config option check, no?

clamattia replied:
I have assumed, that we want to receive `priority tagged VLAN` even if it is disabled.
An alternative approach would be, that it has to be enabled, but the `VLAN_COUNT=0` of actual virtual interfaces is zero. Is this what you have in mind?

Ie. If someone whats to receive `priority tagged vlan` headers, they need to enable `CONFIG_NET_VLAN` but they will not have any actual virtual LAN interface?

clamattia replied:
As mentioned above. I reverted this logic. User must now enable `CONFIG_NET_VLAN` to receive priority tagged frames.

clamattia commented:
Idea is now, that configuration to receive only priority tagged frames without actually enabling a virtual interface looks like:
```Kconfig
CONFIG_NET_VLAN=y
NET_VLAN_COUNT=0
```
I will have to test, if this actually works.

Feedback is welcome :)

clamattia commented on include/zephyr/net/ethernet.h:
```diff
@@ -117,11 +117,7 @@ struct net_eth_addr {
 #define _NET_ETH_MAX_HDR_SIZE		(sizeof(struct net_eth_vlan_hdr))
-#else
-#define _NET_ETH_MAX_HDR_SIZE		(sizeof(struct net_eth_hdr))
-#endif
```
I reverted this. Users, who want to receive priority tagged frames need to enable `CONFIG_NET_VLAN` now.

clamattia commented:
Testing it leads to a problem.
With `CONFIG_NET_VLAN_COUNT=1`:
```
uart:~$ net iface
Hostname: xxx


Interface eth0 (0x20001ac0) (Ethernet) [1]
===================================
Link addr : 00:80:E1:96:0E:73
MTU       : 1500
Flags     : AUTO_START,IPv4
Device    : ethernet@40028000 (0x8076c44)
Ethernet capabilities supported:
        Virtual LAN
        10 Mbits
        100 Mbits
IPv4 unicast addresses (max 1):
        <none>
IPv4 multicast addresses (max 2):
        224.0.0.1
IPv4 gateway : 0.0.0.0

Interface net0 (0x20001ba4) (Virtual) [2]
==================================
```
With `CONFIG_NET_VLAN_COUNT=0` gives:
```
uart:~$ net iface
Hostname: xxx


Interface eth0 (0x20001ab8) (Ethernet) [1]
===================================
Link addr : 00:80:E1:96:0E:73
MTU       : 1500
Flags     : AUTO_START,IPv4
Device    : ethernet@40028000 (0x8076780)
Ethernet capabilities supported:
	Virtual LAN
	10 Mbits
	100 Mbits
IPv4 not enabled for this interface.
```

Why does it say `Ipv4 not enabled` ? Probably `CONFIG_NET_VLAN_COUNT=0` is breaking an assumption?

jukkar commented:
I am not really sure if my understanding of this feature is correct, but I thought about VLAN 0 like this:
* The `CONFIG_NET_VLAN` must be enabled to receive VLAN frames. If VLAN is disabled, then user is not interested in VLAN support so VLAN 0 support is not needed either.
* If we receive a VLAN frame with tag value 0, then the packet is fed into the underlying network interface instead of feeding it into VLAN interface.
* Normal VLAN traffic (tags > 0) are placed to the proper VLAN interface
* The `CONFIG_NET_VLAN_COUNT` is handled normally, IMHO it should have min value of 1 (because if 0, then basically VLAN is disabled which is not a point here).

WDYT?

clamattia commented:
>     * The `CONFIG_NET_VLAN` must be enabled to receive VLAN frames. If VLAN is disabled, then user is not interested in VLAN support so VLAN 0 support is not needed either.
Fair enough. My initial draft wanted to not require the user to select `CONFIG_NET_VLAN` to have support for VLAN 0. But I can get down with this decision.

>     * If we receive a VLAN frame with tag value 0, then the packet is fed into the underlying network interface instead of feeding it into VLAN interface.
Agree.

>     * Normal VLAN traffic (tags > 0) are placed to the proper VLAN interface
Agree.

>     * The `CONFIG_NET_VLAN_COUNT` is handled normally, IMHO it should have min value of 1 (because if 0, then basically VLAN is disabled which is not a point here).
Disagree. How else is the user supposed to express their interest in selecting VLAN 0 without having any virtual interface for normal VLAN traffic (tags > 0). This is my usecase. I.E. They want VLAN 0 traffic but no actual proper virtual interface.

clamattia commented on subsys/net/l2/ethernet/Kconfig:
```diff
@@ -41,7 +41,7 @@ config NET_VLAN
 	int "Max VLAN tags supported in the system"
 	default 1
-	range 1 $(UINT8_MAX)
+	range 0 $(UINT8_MAX)
```
See my explenation: https://github.com/zephyrproject-rtos/zephyr/pull/83734#issuecomment-2589953895

clamattia replied:
I try to find a way to make room for the use case of receiving VLAN 0 frames without having an actual virtual interface enabled. This is my use case.

jukkar replied:
Hmm, I am not sure you need to enable the VLAN virtual interface. I mean the enabled indicates that the interface has a tag assigned to it. But to receive tag 0, this is not needed.

clamattia replied:
> Hmm, I am not sure you need to enable the VLAN virtual interface. I mean the enabled indicates that the interface has a tag assigned to it. But to receive tag 0, this is not needed.

Are you saying, that I should go back to initial draft of always (without `CONFIG_VLAN` enabled) forwarding tag 0 to the native interface?

This would be my preferred solution. But it has other consequences. It requires, the increase in the expected header size as in the initial draft for example.

jukkar commented:
> They want VLAN 0 traffic but no actual proper virtual interface.

I don't think this is really feasible without major changes (not fully sure about this). Fortunately you do not need to enable the virtual interface to support VLAN 0.

clamattia commented:
> > They want VLAN 0 traffic but no actual proper virtual interface.
> 
> I don't think this is really feasible without major changes (not fully sure about this). Fortunately you do not need to enable the virtual interface to support VLAN 0.

Ok, will revert to initial draft about this respect. 👍

clamattia commented:
I have reverted this aspect. Now everyone will be able to receive vlan frames with id `0` on their native interface. No need to enable `CONFIG_VLAN` for this.

clamattia commented on include/zephyr/net/ethernet.h:
```diff
@@ -117,11 +117,12 @@ struct net_eth_addr {
+ * Leave enough space in the header for vlan, even if CONFIG_VLAN is not enabled,
+ * so that frames with a vlan-header and vlan-id NET_VLAN_ID_NATIVE can be
+ * forwarded to the native interface.
+ */
 #define _NET_ETH_MAX_HDR_SIZE		(sizeof(struct net_eth_vlan_hdr))
-#else
-#define _NET_ETH_MAX_HDR_SIZE		(sizeof(struct net_eth_hdr))
-#endif
```
> Hmm, I am not sure you need to enable the VLAN virtual interface. I mean the enabled indicates that the interface has a tag assigned to it. But to receive tag 0, this is not needed.

It is needed. Otherwise id 0 frames can not be received.

I can try to fix, but I would like to get your approval on the approach first.

clamattia replied:
reverted and created separate issue: https://github.com/zephyrproject-rtos/zephyr/issues/84023

clamattia commented:
As discussed in discord, have reverted to more basic solution and added issue for optimized solution here: https://github.com/zephyrproject-rtos/zephyr/issues/84023

clamattia commented:
I rebased, please re-review @jukkar 

Also, ideally, we would find a solution for https://github.com/zephyrproject-rtos/zephyr/issues/84023

jukkar committed:
net: ethernet: Properly handle VLAN tag 0

Packets are forwarded to the native interface or in other words,
the vlan header is simply stripped and ignored. This feature is called
'priority tagging'.

Signed-off-by: Cla Mattia Galliard <cla-mattia.galliard@zuehlke.com>
Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
I pushed my proposal over the original PR. It has changes from #84158 in order to make testing possible. The actual VLAN tag 0 handling can be found in the last commit https://github.com/zephyrproject-rtos/zephyr/pull/83734/commits/3ce3777b55471b17ba63da69deb20085efd743aa

jukkar committed:
tests: net: vlan: Add VLAN tag 0 test

Make sure we are able to receive VLAN tag 0 packet, and
verify that reply packet is sent correctly.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>

jukkar commented:
* Added tests for VLAN tag 0
* Rebased to latest main
